esp32-srne-asf-example.yaml

```
@@ -0,0 +1,827 @@
+    filters:
+      - multiply: 0.1
+
+  - platform: modbus_controller
+    modbus_controller_id: smg0
+    name: "MPPT 1 Power"
+    id: mppt1_power
+    address: 0x109
+    unit_of_measurement: "W"
+    register_type: holding
+    value_type: U_WORD
+    device_class: Power
+    accuracy_decimals: 1
+
+  - platform: modbus_controller
+    modbus_controller_id: smg0
+    name: "Charge Power"
+    id: charge_power
+    address: 0x10E
+    unit_of_measurement: "W"
+    register_type: holding
+    value_type: U_WORD
+    accuracy_decimals: 1
+
+  - platform: modbus_controller
+    modbus_controller_id: smg0
+    name: "MPPT 2 Voltage"
+    id: mppt2_voltage
+    address: 0x10F
+    unit_of_measurement: "V"
+    register_type: holding
+    value_type: U_WORD
+    accuracy_decimals: 1
+    filters:
+      - multiply: 0.1
+
+  - platform: modbus_controller
+    modbus_controller_id: smg0
+    name: "MPPT 2 Current"
+    id: mppt2_current
+    address: 0x110
+    unit_of_measurement: "A"
+    register_type: holding
+    value_type: U_WORD
+    accuracy_decimals: 1
+    filters:
+      - multiply: 0.1
+
+  - platform: modbus_controller
+    modbus_controller_id: smg0
+    name: "MPPT 2 Power"
+    id: mppt2_power
+    address: 0x111
+    unit_of_measurement: "W"
+    register_type: holding
+    value_type: U_WORD
+    device_class: Power
+    accuracy_decimals: 1
+
+  - platform: template
+    name: "Solar Power"
+    id: solar_power
+    unit_of_measurement: "W"
+    accuracy_decimals: 0
+    device_class: Power
+    update_interval: 1s
+    lambda: |-
+      return (id(mppt1_power).state + id(mppt2_power).state);
+
+  - platform: modbus_controller
+    modbus_controller_id: smg0
+    id: battery_rated_voltage
+    name: "Rated Voltage"
+    address: 0x204
+    unit_of_measurement: "V"
+    register_type: holding
+    value_type: U_WORD
+    accuracy_decimals: 1
+    filters:
+      - multiply: 0.01
+
+#  - platform: modbus_controller
+#    modbus_controller_id: smg0
+#    id: time_year
+#    name: "Time Year"
+#    address: 0x020C
+#    register_type: holding
+#    bitmask: 0xFF00
+#    value_type: U_WORD
+
+  - platform: modbus_controller
+    modbus_controller_id: smg0
+    id: grid_voltage
+    name: "Grid Voltage"
+    address: 0x213
+    unit_of_measurement: "V"
+    register_type: holding
+    value_type: U_WORD
+    accuracy_decimals: 1
+    filters:
+      - multiply: 0.1
+
+  - platform: modbus_controller
+    modbus_controller_id: smg0
+    id: grid_current_a
+    name: "Grid Current A"
+    address: 0x214
+    unit_of_measurement: "A"
+    register_type: holding
+    value_type: U_WORD
+    accuracy_decimals: 1
+    filters:
+      - multiply: 0.1
+
+  - platform: modbus_controller
+    modbus_controller_id: smg0
+    id: grid_frequency
+    name: "Grid Frequency"
+    address: 0x215
+    unit_of_measurement: "Hz"
+    register_type: holding
+    value_type: U_WORD
+    accuracy_decimals: 1
+    filters:
+      - multiply: 0.01
+
```

I believe you are missing these two important registers, which are the inverter load voltage (phase A) and load current (phase A).

My inverter has only one phase, so only phase A reads data.

```suggestion

  - platform: modbus_controller
    modbus_controller_id: srne_0
    id: inverter_voltage
    name: "Inverter Voltage"
    address: 0x216
    unit_of_measurement: "V"
    register_type: holding
    value_type: U_WORD
    accuracy_decimals: 1
    filters:
      - multiply: 0.1

  - platform: modbus_controller
    modbus_controller_id: srne_0
    id: inverter_current
    name: "Inverter Current"
    address: 0x219
    unit_of_measurement: "A"
    register_type: holding
    value_type: U_WORD
    accuracy_decimals: 1
    filters:
      - multiply: 0.1
```

Even though [protocol 3.9](https://github.com/HotNoob/PythonProtocolGateway/blob/main/protocols/srne/srne_v3.9.holding_registry_map.csv) doesn't mention them, I suppose they should work, given the 0x218 (inverter frequency) works, which seems also not part of the 3.9 manual.

---

Added.